KarhooUISDK/Workers/Loyalty/LoyaltyWorker.swift

```diff
@@ -0,0 +1,292 @@
+    private let loyaltyService: LoyaltyService
+    private let loyaltyPreAuthWorker: LoyaltyPreAuthWorker
+    private let analytics: Analytics
+
+    // MARK: - Properties
+
+    // MARK: Internal properties
+
+    static let shared = KarhooLoyaltyWorker()
+
+    var isLoyaltyEnabled: Bool { getIsLoyaltyEnabled() }
+    var modelSubject = CurrentValueSubject<Result<LoyaltyUIModel?>, Never>(.success(result: nil))
+    var modeSubject = CurrentValueSubject<LoyaltyMode, Never>(.none)
+
+    // MARK: Private properties
+
+    private var canBurnSubject = CurrentValueSubject<Bool, Never>(false)
+    private var canEarnSubject = CurrentValueSubject<Bool, Never>(false)
+    private var earnPointsSubject = CurrentValueSubject<Int?, Never>(nil)
+    private var burnPointsSubject = CurrentValueSubject<Int?, Never>(nil)
+    private var currentBalanceSubject = CurrentValueSubject<Int?, Never>(nil)
+
+    private var quote: Quote?
+    /// Error returned by `getBurnPoints` method, indicating the burn is not possible. It should be used to update PreAuth worker.
+    private var burnError: KarhooError?
+    private var cancellables: Set<AnyCancellable> = []
+
+    // MARK: - Lifecycle
+
+    init(
+        userService: UserService = Karhoo.getUserService(),
+        loyaltyService: LoyaltyService = Karhoo.getLoyaltyService(),
+        loyaltyPreAuthWorker: LoyaltyPreAuthWorker = KarhooLoyaltyPreAuthWorker(),
+        analytics: Analytics = KarhooUISDKConfigurationProvider.configuration.analytics()
+    ) {
+        self.userService = userService
+        self.loyaltyService = loyaltyService
+        self.loyaltyPreAuthWorker = loyaltyPreAuthWorker
+        self.analytics = analytics
+
+        setupPublishers()
+    }
+    // MARK: - Endpoints
+
+    func setup(using quote: Quote) {
+        self.quote = quote
+        getData()
+    }
+
+    func getLoyaltyNonce(completion: @escaping (Result<LoyaltyNonce>) -> Void) {
+        guard isLoyaltyEnabled else {
+            completion(.failure(error: KarhooLoyaltyError.none))
+            return
+        }
+        loyaltyPreAuthWorker.getLoyaltyPreAuthNonce(completion: completion)
+    }
+
+    // MARK: - Private methods
+
+    private func getData() {
+        guard isLoyaltyEnabled else { return }
+        getLoyaltyStatus()
+        getEarnedPoints()
+        getBurnedPoints()
+    }
+
+    private func setupPublishers() {
+        /// Every time any of those values changes, the sink closure is triggered, and, as a result, the model and PreAuthWorker are updated.
+        let zippedEarnBurn = Publishers.CombineLatest4(
+            earnPointsSubject,
+            burnPointsSubject,
+            canEarnSubject,
+            canBurnSubject
+        )
+        Publishers.CombineLatest3(
+            currentBalanceSubject,
+            zippedEarnBurn,
+            modeSubject
+        )
+        .sink(receiveValue: { [weak self] values in
+            let earnBurnValues = values.1
```

Comment:
Is there a way to label these values so that we avoid using `values.1`, `values.2`, etc?

Reply:
There is no such way. We create a custom publisher with tuple output type. The best we can do is to create readable identifiers as soon as value is obtained.